activerecord/lib/active_record/counter_cache.rb

```diff
@@ -66,7 +66,11 @@ def reset_counters(id, *counters, touch: nil)
           updates.merge!(touch_updates)
         end
 
-        unscoped.where(primary_key => object.id).update_all(updates) if updates.any?
+        if updates.any?
+          id = object.id
+          id = [id] if composite_primary_key? && id.is_a?(Array) && !id[0].is_a?(Array)
```

Comment:
I think we're way past the point where we need to take a step back and figure out how to not have to check for CPK in such a ad hoc way everywhere.

Even if we ignore how ugly this is, and fix all the call sites we may have missed, we're likely to constantly break CPK support when adding new features.

Reply:
Generally we expected that `where(primary_key => id)` will "just work", or at least `where(primary_key => [id])` will work regardless of what shape `id` has.
But it looks like we end up in a situation where only "multi" `where(primary_key => [id])` where `[id]` is an array of arrays for composite primary key models and array of plan ids for regular models.


One of the possible options that was considered is to introduce some kind of a `Model.primary_key_scope(id_or_ids)` helper (at least for internal use) which will always know how to build a correct relation. The helper implementation may end up having similar `if` checks but at least consumers won't need to bother.

It may be a bit harder but in the long term if we had `primary_key_scope` concept we could make sure that CPK relations and regular model relations have different implementation of the method so there will be no branching

Reply:
Hey @fatkodima, do you remember why this line needed the array wrap as well?

In this case we know that `id = object.id` will always represent one identifier so ` unscoped.where(primary_key => [id])` should just work. I tried it quickly and none of the tests seem to fail

The `update_counters` does need at least something as indeed we can't know whether passed `id` is an array of ids or just a composite primary key array

Reply:
Yeah, correct 🤦 We had `object.id`, not just `id`.

Reply:
I've come across this when implementing composite primary keys in the [positioning](https://github.com/brendon/positioning) gem. The format for a composite key (needing to be an array wrapped in an array) was surprising but as you say @nvasilevski, it does work fine. At least it means the codebase doesn't care about composite primary keys in terms of implementation detail. Is there a reason for the double-wrapping requirement? And was there any further progress on a `primary_key_scope`?